### PR TITLE
[improve][misc] Use shared Jackson ObjectMapper to reduce overhead and remove ThreadLocal solution

### PR DESCRIPTION
**Pull request by lhotari**
### Motivation

Creating a Jackson ObjectMapper is an expensive operation. A  thread local ObjectMapper instance is already used in most locations.

Additional motivation is to be able to use a properly configured ObjectMapper instance with Jackson Java 8 modules. That PR is #19161 .

### Modifications

* Make use of shared ObjectMapper instances. 
* Remove the ThreadLocal ObjectMapper instances and replace with a global singleton instance
  * In some old versions of Jackson there was a performance benefit of using a ThreadLocal instance. This isn't necessary with the Jackson version that is used.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/lhotari/pulsar/pull/124

### CODE REVIEW

**Review by nicoloboschi — APPROVED** [review]

**Review by HQebupt — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericJsonWriter.java [hunk 1]
```diff
@@ -18,24 +18,21 @@
...
 import org.apache.pulsar.client.api.SchemaSerializationException;
 import org.apache.pulsar.client.api.schema.GenericRecord;
 import org.apache.pulsar.client.api.schema.SchemaWriter;
+import org.apache.pulsar.common.util.ObjectMapperFactory;
 
 public class GenericJsonWriter implements SchemaWriter<GenericRecord> {
 
-    private final ObjectMapper objectMapper;
-
     public GenericJsonWriter() {
```
Comment (cbornet) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericJsonWriter.java:29]: Remove this constructor ?

**Review by nicoloboschi — COMMENTED** [review]
@lhotari I see the improvement here, a couple of comments:
- wouldn't be better to only expose the read/write method + `unsafeGetInstance` ? My concern is that you could call `ObjectMapperFactory.getInstance().setDateFormat(xx)` changing the internal configuration 
- `ObjectMapperFactory` naming should matches the singleton pattern

**Comment by lhotari** [comment]
> @lhotari I see the improvement here, a couple of comments:
> 
> * wouldn't be better to only expose the read/write method + `unsafeGetInstance` ? My concern is that you could call `ObjectMapperFactory.getInstance().setDateFormat(xx)` changing the internal configuration

You are right. I assume you mean returning an ObjectReader/ObjectWriter (which are immutable) instead of returning a ObjectMapper at all. It would be a better approach.
One detail is that the current threadlocal solution is already "unsafe".

> * `ObjectMapperFactory` naming should matches the singleton pattern

There's a need to have different type of ObjectMapper instances already. I noticed that replacing `new ObjectMapper()` with the shared instance broke things since the serialization config if different (`Include.ALWAYS` is the default, but the configured one has `Include.NON_NULL`). Therefore the name `ObjectMapperFactory` could be justified although it's not an actual Factory. 
What name would you suggest?

**Comment by lhotari** [comment]
@nicoloboschi PTAL . I have added a ProtectedObjectMapper class which prevents mutating the shared ObjectMapper config.

**Review by nicoloboschi — APPROVED** [review]
LGTM 
I like the new pattern 👍

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/LocalPoliciesResources.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ public CompletableFuture<Optional<CacheGetResult<LocalPolicies>>> getLocalPolici
     public CompletableFuture<Void> setLocalPoliciesWithVersion(NamespaceName ns, LocalPolicies policies,
                                                                Optional<Long> version) {
         try {
-            byte[] content = ObjectMapperFactory.getThreadLocal().writeValueAsBytes(policies);
+            byte[] content = ObjectMapperFactory.getMapper().writer().writeValueAsBytes(policies);
```
Comment (cbornet) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/LocalPoliciesResources.java:69]: How about having `ObjectMapperFactory.getWriter()` and `ObjectMapperFactory.getReader()` methods ?

Comment (lhotari) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/LocalPoliciesResources.java:69]: I was thinking of that, but the challenge is that there's also the Yaml versions and the ObjectMapper configured to filter null fields. That would result in a lot of similar methods. It's a tradeoff. 
Leaving ObjectMapperFactory as-is and creating a completely new class would be another option. I'd suggest that it's left as another PR to continue refactoring.

**Pull request merged**